sig-scalability/slos/network_programming_latency.md

```diff
@@ -0,0 +1,95 @@
+    in 99% of programmers (e.g. iptables). That requires tracking metrics on
+    per-change base (which we can't do efficiently).
+  - we admit that the SLO is a bit weaker in that form (i.e. it doesn't necessary
+    force that a given change is reflected in 99% of programmers with a given
+		99th percentile latency), but it's close enough approximation.
+
+### How to measure the SLI.
+The method of measuring this SLI is not obvious, so for completeness we describe
+it here how it will be implemented with all caveats.
+1. We assume that for the in-cluster load-balancing programming we are using
+Kubernetes `Endpoints` objects.
+1. We will introduce a dedicated annotation for `Endpoints` object (name TBD).
+1. Endpoints controller (while updating a given `Endpoints` object) will be
+setting value of that annotation to the timestamp of the change that triggered
+this update:
+- for pod transition between `Ready` and `NotReady` states, its timestamp is
+  simply part of pod condition
+- TBD for service updates (ideally we will add `LastUpdateTimestamp` field in
+  object metadata next to already existing `CreationTimestamp`. The data is
+  already present at storage layer, so it won't be hard to propagate that.
+1. The in-cluster load-balancing programmer will export a prometheus metric
+once done with programming. The latency of the operation is defined as
+difference betweem timestamp of then whe operation is done and timestamp
+recorded in the newly introduced annotation.
+
+#### Caveats
+There are a couple of caveats to that measurement method:
+1. Single `Endpoints` object may batch multiple pod state transition. <br/>
+In that case, we simply choose the oldest one (and not expose all timestamps
+to avoid theoretically unbounded growth of the object). That makes the metric
+imprecise, but the batching period should be relatively small comparing
+to whole end-to-end flow.
+1. A single pod may transition its state multiple times within batching
+period. <br/>
+For that case, we will add additional cache in Endpoints controller caching
+the first observed transition timestamp for each pod. The cache will be
+cleared when controller picks up a pod into Endpoints object update. This is
+consistent with choosing the oldest update in the above point. <br/>
+Initially, we may consider simply ignoring this fact.
+1. Components may fall out of watch window history and thus miss some watch
+events. <br/>
+This may be the case for both Endpoints controller or kube-proxy (or other
+network programmers if used instead). That becomes a problem when a single
+object changed multiple times in the meantime (otherwise informers will
+deliver handlers on relisting). Additionally, this can happen only when
+components are too slow in processing events (that would already be reflected
+in metrics) or (sometimes) after kube-apiserver restart. Given that, we are
+going to neglect this problem to avoid unnecessary complications for little
+or no gain.
+
+**TODO:**
```

Comment:
Is there more coming here?

Reply:
Not that I'm aware of now - I should probably remove it.

Reply:
Just removed that.